dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -1659,9 +1659,20 @@ consolidateRecordLiteral = concatMap adapt . Map.toList
 escapeSingleQuotedText :: Text -> Text
 escapeSingleQuotedText inputText = outputText
   where
-    outputText = substitute "${" "''${" (substitute "''" "'''" inputText)
+    outputText = go "" inputText
+
+    go :: Text -> Text -> Text
+    go acc xs
+        | xs == "" = acc
+        | otherwise = let (ys, xs') = f xs in go (acc <> ys) xs'
+
+    f xs
+        | Just xs' <- "'${" `Text.stripPrefix` xs = ("${\"'\"}''${", xs')
```

Comment:
I don't think we should produce a formatted string interpolation in a `Text -> Text` transformation. I think it would be better to include do this within `multilineChunks`, so we can produce proper `Chunks` as the output..

Reply:
@sjakobi  You're right. I moved the escaping to function `escapeMultilineChunks` in the most recent commit.
I merged `escapeSingleQuotedText` and `escapeControlCharacters` into that function as well.